tianshou/highlevel/agent.py

```diff
@@ -0,0 +1,598 @@
+import logging
+import typing
+from abc import ABC, abstractmethod
+from typing import Any, Generic, TypeVar, cast
+
+import gymnasium
+from sensai.util.string import ToStringMixin
+
+from tianshou.data import Collector, ReplayBuffer, VectorReplayBuffer
+from tianshou.highlevel.config import SamplingConfig
+from tianshou.highlevel.env import Environments
+from tianshou.highlevel.module.actor import (
+    ActorFactory,
+)
+from tianshou.highlevel.module.core import (
+    ModuleFactory,
+    TDevice,
+)
+from tianshou.highlevel.module.critic import CriticEnsembleFactory, CriticFactory
+from tianshou.highlevel.module.module_opt import (
+    ActorCriticModuleOpt,
+)
+from tianshou.highlevel.optim import OptimizerFactory
+from tianshou.highlevel.params.policy_params import (
+    A2CParams,
+    DDPGParams,
+    DiscreteSACParams,
+    DQNParams,
+    IQNParams,
+    NPGParams,
+    Params,
+    ParamsMixinActorAndDualCritics,
+    ParamsMixinLearningRateWithScheduler,
+    ParamTransformerData,
+    PGParams,
+    PPOParams,
+    REDQParams,
+    SACParams,
+    TD3Params,
+    TRPOParams,
+)
+from tianshou.highlevel.params.policy_wrapper import PolicyWrapperFactory
+from tianshou.highlevel.persistence import PolicyPersistence
+from tianshou.highlevel.trainer import TrainerCallbacks, TrainingContext
+from tianshou.highlevel.world import World
+from tianshou.policy import (
+    A2CPolicy,
+    BasePolicy,
+    DDPGPolicy,
+    DiscreteSACPolicy,
+    DQNPolicy,
+    IQNPolicy,
+    NPGPolicy,
+    PGPolicy,
+    PPOPolicy,
+    REDQPolicy,
+    SACPolicy,
+    TD3Policy,
+    TRPOPolicy,
+)
+from tianshou.trainer import BaseTrainer, OffpolicyTrainer, OnpolicyTrainer
+from tianshou.utils.net.common import ActorCritic
+
+CHECKPOINT_DICT_KEY_MODEL = "model"
+CHECKPOINT_DICT_KEY_OBS_RMS = "obs_rms"
+TParams = TypeVar("TParams", bound=Params)
+TActorCriticParams = TypeVar(
+    "TActorCriticParams",
+    bound=Params | ParamsMixinLearningRateWithScheduler,
+)
+TActorDualCriticsParams = TypeVar(
+    "TActorDualCriticsParams",
+    bound=Params | ParamsMixinActorAndDualCritics,
+)
+TDiscreteCriticOnlyParams = TypeVar(
+    "TDiscreteCriticOnlyParams",
+    bound=Params | ParamsMixinLearningRateWithScheduler,
+)
+TPolicy = TypeVar("TPolicy", bound=BasePolicy)
+log = logging.getLogger(__name__)
+
+
+class AgentFactory(ABC, ToStringMixin):
```

Comment:
I don't think it provides any benefit. Forcing kwargs makes sense only if there is a danger getting parameters mixed up otherwise, which is the case if there is a large number of parameters or if there are several parameters of the same type (that don't have a natural ordering).
Python is not Objective-C and we shouldn't enforce such paradigms if there isn't an obvious benefit. It would likely irritate users.